worlds/jakanddaxter/Regions.py

```diff
@@ -0,0 +1,210 @@
+
+    # Define the interconnecting rules.
+    menu.connect(gr)
+    gr.connect(sv)  # Geyser Rock modified to let you leave at any time.
+    sv.connect(fj)
+    sv.connect(sb)
+    sv.connect(mi, rule=lambda state: state.has("Fisherman's Boat", player))
+    sv.connect(fc, rule=lambda state: state.has("Power Cell", player, fc_count))  # Normally 20.
+    fc.connect(rv)
+    rv.connect(pb)
+    rv.connect(lpc)
+    rvp.connect(bs)  # rv->rvp/rvc connections defined internally by RockVillageRegions.
+    rvc.connect(mp, rule=lambda state: state.has("Power Cell", player, mp_count))  # Normally 45.
+    mpr.connect(vc)  # mp->mpr connection defined internally by MountainPassRegions.
+    vc.connect(sc)
+    vc.connect(sm, rule=lambda state: state.has("Snowy Mountain Gondola", player))
+    vc.connect(lt, rule=lambda state: state.has("Power Cell", player, lt_count))  # Normally 72.
+    lt.connect(gmc)  # gmc->fb connection defined internally by GolAndMaiasCitadelRegions.
+
+    # Set the completion condition.
+    if options.jak_completion_condition == CompletionCondition.option_cross_fire_canyon:
+        multiworld.completion_condition[player] = lambda state: state.can_reach(rv, "Region", player)
+
+    elif options.jak_completion_condition == CompletionCondition.option_cross_mountain_pass:
+        multiworld.completion_condition[player] = lambda state: state.can_reach(vc, "Region", player)
+
+    elif options.jak_completion_condition == CompletionCondition.option_cross_lava_tube:
+        multiworld.completion_condition[player] = lambda state: state.can_reach(gmc, "Region", player)
+
+    elif options.jak_completion_condition == CompletionCondition.option_defeat_dark_eco_plant:
+        multiworld.completion_condition[player] = lambda state: state.can_reach(fjp, "Region", player)
+
+    elif options.jak_completion_condition == CompletionCondition.option_defeat_klaww:
+        multiworld.completion_condition[player] = lambda state: state.can_reach(mp, "Region", player)
+
+    elif options.jak_completion_condition == CompletionCondition.option_defeat_gol_and_maia:
+        multiworld.completion_condition[player] = lambda state: state.can_reach(fb, "Region", player)
+
+    elif options.jak_completion_condition == CompletionCondition.option_open_100_cell_door:
+        multiworld.completion_condition[player] = lambda state: state.can_reach(fd, "Region", player)
+
+    # As a final sanity check on these options, verify that we have enough locations to allow us to cross
+    # the connector levels. E.g. if you set Fire Canyon count to 99, we may not have 99 Locations in hub 1.
+    verify_connector_level_accessibility(multiworld, options, player)
+
+    # Also verify that we didn't overload the trade amounts with more orbs than exist in the world.
+    verify_orbs_for_trades(multiworld, options, player)
+
+
+def verify_connector_level_accessibility(multiworld: MultiWorld, options: JakAndDaxterOptions, player: int):
```

Comment:
> this is especially true in a multiworld and I am wondering a bit if this function should only be run in soloworlds ... What prompted its creation, out of curiosity, just worlds failing in fill?

Yeah, solo world generation was the main impetus for this. I can't remember if I tested this function with multiple worlds to see if it respected the larger initial location pool, but I suppose that was my hope when I passed `multiworld` to the CollectionState constructor. 

If I used a lookup table that mapped all possible option combinations to a number of locations, I was worried that future changes to Regions/Rules/Options would result in a stale table and defeat the purpose of the function. As a programmer I do have a habit of overbuilding things, and in this instance I was trying to ensure we had one source of truth.

Reply:
You could just run this code separately (like, in a different branch), use that to count the orb counts and then just populate a table, and you can update that as regions change. You could even have tests set up to verify the values change when they should. But idk, these functions are pretty strange to me and feel like a very heavy-handed solution to a problem that shouldn't need this kind of solution.

A thing that's confusing me, even _if_ move rando is enabled, the generator can just give you all of your moves right from the start, so there should be a pretty clear maximum number of orbs each gate could cost (locations gettable with all relevant items - number of relevant items). I don't think that would be a _fun_ world, but if the goal here is to prevent fill from failing to find solutions to highly-restricted starts, you may be better off just placing progression items yourself or even into start_inventory. I'll try to think through this some more when I have time and see if I can come up with a simpler or less expensive solution.

Reply:
Honestly, the problem is a bit complicated though, so it might be best to just drop a big description/explanation in #ap-world-dev in the Discord and see what people suggest.

Reply:
[For posterity, because it was a _great_ discussion.](https://discord.com/channels/731205301247803413/1214608557077700720/1276332997213290496)